File: problem_5/problem_5.py
# 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
# What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

import time

def is_prime(n):
    """ Returns Boolean """
    if n <= 1: return False
    if n == 2 or n == 3: return True
    if n % 2 == 0 or n % 3 == 0: return False
    i, w, = 5, 2
    while i * i <= n:
        if n % i == 0:
            return False
        i += w
        w = 6 - w
    return True

def prime_factorization(n):
    """ 
    Assumes n >= 2
    Returns a dict mapping the prime factors of n and their respective powers 
    """
    if is_prime(n):
        return {n: 1}
    prime_factors = {2:0, 3:0,}
    while n % 2 == 0:
        prime_factors[2] += 1
        n /= 2
    while n % 3 == 0:
        prime_factors[3] += 1
        n /= 3  
    for i in range(5, int(n**0.5)+1, 2):
        if not is_prime(i):
            continue
        while n % i == 0:
            try:
                prime_factors[i] += 1
            except KeyError:
                prime_factors[i] = 1
            n /= i
    return prime_factors        

def problem_5(n):
    """ Returns the smallest number divisible by every number <= n """
    output = 1
    prime_factor_map = {}

    for i in range(2, n+1):
        prime_factors = prime_factorization(i)
        for p,e in prime_factors.items():
            try:
                if prime_factor_map[p] < e:
                    prime_factor_map[p] = e
            except KeyError:
                prime_factor_map[p] = e

    for p,e in prime_factor_map.items():
        output *= pow(p,e)
    return output

start = time.time()
solution = problem_5(20)
print(f"{solution} found in {time.time() - start} seconds.")
# 232792560 found in 3.1948089599609375e-05 seconds.

File: problem_5/problem_5.js
// # 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
// # What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

const now = require("performance-now");

const isPrime = n => {
    if (n <= 1){ return false; }
    if (n == 2 || n == 3){ return true; }
    if (n % 2 === 0 || n % 3 === 0){ return false; }

    let i = 5, w = 2;
    while (i * i <= n){
        if (n % i === 0){ return false; }
        i += w;
        w = 6 - w;
    }
    return true;
}

const primeFactorization = n => {
    // Assumes n >= 2
    // Returns a dict mapping the prime factors of n and their respective powers 
    if (isPrime(n)){ 
        let primeFactors = {}
        primeFactors[n] = 1;
        return primeFactors;
    }
    let primeFactors = { 2:0, 3:0 }
    while (n % 2 === 0){
        primeFactors[2] += 1;
        n /= 2;
    }
    while (n % 3 === 0){
        primeFactors[3] += 1;
        n /= 3;
    }
    for (var i = 5; i < parseInt(Math.sqrt(n) + 1, 10); i+= 2){
        if (!isPrime(i)){ continue }
        while (n % i === 0){
            if (!primeFactors[i]){ primeFactors[i] = 1; }
            else { primeFactors[i] += 1; }
            n /= i;
        }
    }
    return primeFactors;
}

const problem_5 = n => {
    //  Returns the smallest number divisible by every number <= n 
    let output = 1;
    let primeFactorMap = {};

    for (var i = 2; i < n + 1; i++){
        let primeFactors = primeFactorization(i);
        for (const [p,e] of Object.entries(primeFactors)){
            if (!primeFactorMap[p]){ primeFactorMap[p] = e; }
            else if (primeFactorMap[p] < e){ primeFactorMap[p] = e; }
        }
    }
    for (const [p,e] of Object.entries(primeFactorMap)){
        output *= parseInt(p, 10) ** e;
    }
    return output
}

let start = now();
let solution = problem_5(20)
console.log(`${solution} found in ${(now() - start) / 1000} seconds.`)
// 232792560 found in 0.00023718999999999824 seconds.